<a href="https://www.kaggle.com/code/ayushs9020/understanding-the-competition-commonlit?scriptVersionId=136782207" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#006600; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #003300">Common Lit Grading for Students</p>

<div style="border-radius:10px; border:#DEB887 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
    
<img src = "https://pressbooks.howardcc.edu/app/uploads/sites/63/2021/01/balloon-223x300.png">

The $CommonLit$ - $Evaluate$ $Student$ $Summaries$ competition is a Kaggle competition that `challenges participants` to build a model that can `automatically assess the quality of summaries` written by students in grades 3-12. The goal of the competition is to help `teachers` and `learning platforms provide better feedback` to students on their writing.
    
<img src = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQhAQr_CLNzbbHFRgbxGee3nvg24v-LdkmQ8fMJfG0YyA9zOvCpr8zq8nFlJz9vh-ThI9I&usqp=CAU">

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#00FFFF; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #00FFFF">1 | Advisory🚀</p>

<div style="border-radius:10px; border:#00FFFF solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
    
* Train data needs to be merged with to get both summaries and prompts

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#FFA500; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #FFA500">2 | Data 📈</p>

<div style="border-radius:10px; border:#FFA500 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

<img src = "https://miro.medium.com/v2/resize:fit:500/1*ZhYNqU2y96_f3QkWq9oiWQ.jpeg" width = 300>
    
Now lets dive into the data

In [1]:
import pandas as pd

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/roberta/Roberta Train Embeds.npy
/kaggle/input/flan-t5/pytorch/base/4/config.json
/kaggle/input/flan-t5/pytorch/base/4/spiece.model
/kaggle/input/flan-t5/pytorch/base/4/README.md
/kaggle/input/flan-t5/pytorch/base/4/tokenizer.json
/kaggle/input/flan-t5/pytorch/base/4/tf_model.h5
/kaggle/input/flan-t5/pytorch/base/4/tokenizer_config.json
/kaggle/input/flan-t5/pytorch/base/4/pytorch_model.bin
/kaggle/input/flan-t5/pytorch/base/4/model.safetensors
/kaggle/input/flan-t5/pytorch/base/4/special_tokens_map.json
/kaggle/input/flan-t5/pytorch/base/4/.gitattributes
/kaggle/input/flan-t5/pytorch/base/4/flax_model.msgpack
/kaggle/input/flan-t5/pytorch/base/4/generation_config.json
/kaggle/input/commonlit-evaluate-student-summaries/sample_submission.csv
/kaggle/input/commonlit-evaluate-student-summaries/prompts_train.csv
/kaggle/input/commonlit-evaluate-student-summaries/summaries_test.csv
/kaggle/input/commonlit-evaluate-student-summaries/summaries_train.csv
/kaggle/input/commonlit-e

<div style="border-radius:10px; border:#FFA500 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $2.1$ $|$ $Train$
    
We have our training data in two different directories `prompts_train`/`summaries_train`

In [2]:
train_pro = pd.read_csv("/kaggle/input/commonlit-evaluate-student-summaries/prompts_train.csv")
train_pro

prompt_id                                    prompt_question  \
0    39c16e  Summarize at least 3 elements of an ideal trag...   
1    3b9047  In complete sentences, summarize the structure...   
2    814d6b  Summarize how the Third Wave developed over su...   
3    ebad26  Summarize the various ways the factory would u...   

                prompt_title  \
0                 On Tragedy   
1  Egyptian Social Structure   
2             The Third Wave   
3    Excerpt from The Jungle   

                                         prompt_text  
0  Chapter 13 \r\nAs the sequel to what has alrea...  
1  Egyptian society was structured like a pyramid...  
2  Background \r\nThe Third Wave experiment took ...  
3  With one member trimming beef in a cannery, an...

<div style="border-radius:10px; border:#FFA500 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

This file contains the actual text values given by the children as the answers to the prompt

In [3]:
train_sum = pd.read_csv("/kaggle/input/commonlit-evaluate-student-summaries/summaries_train.csv")
train_sum

student_id prompt_id  \
0     000e8c3c7ddb    814d6b   
1     0020ae56ffbf    ebad26   
2     004e978e639e    3b9047   
3     005ab0199905    3b9047   
4     0070c9e7af47    814d6b   
...            ...       ...   
7160  ff7c7e70df07    ebad26   
7161  ffc34d056498    3b9047   
7162  ffd1576d2e1b    3b9047   
7163  ffe4a98093b2    39c16e   
7164  fffbccfd8a08    ebad26   

                                                   text   content   wording  
0     The third wave was an experimentto see how peo...  0.205683  0.380538  
1     They would rub it up with soda to make the sme... -0.548304  0.506755  
2     In Egypt, there were many occupations and soci...  3.128928  4.231226  
3     The highest class was Pharaohs these people we... -0.210614 -0.471415  
4     The Third Wave developed  rapidly because the ...  3.272894  3.219757  
...                                                 ...       ...       ...  
7160  They used all sorts of chemical concoctions to...  0.205683  0.380538  
7161  The lowest classes are slaves and farmers slav... -0.308448  0.048171  
7162             they sorta made people start workin... -1.408180 -0.493603  
7163  An ideal tragety has three elements that make ... -0.393310  0.627128  
7164  The meat would smell sour but the would "rub i...  1.771596  0.547742  

[7165 rows x 5 columns]

<div style="border-radius:10px; border:#FFA500 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $2.2$ $|$ $Test$
    
Same as the train, we have our `test` directories also distributed

In [4]:
test_pro = pd.read_csv("/kaggle/input/commonlit-evaluate-student-summaries/prompts_test.csv")
test_pro

prompt_id prompt_question     prompt_title       prompt_text
0    abc123    Summarize...  Example Title 1  Heading\nText...
1    def789    Summarize...  Example Title 2  Heading\nText...

In [5]:
test_sum = pd.read_csv("/kaggle/input/commonlit-evaluate-student-summaries/summaries_test.csv")
test_sum

student_id prompt_id            text
0  000000ffffff    abc123  Example text 1
1  111111eeeeee    def789  Example text 2
2  222222cccccc    abc123  Example text 3
3  333333dddddd    def789  Example text 4

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#800080; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #800080">3 | Tokenization + Preprocessing 💪</p>

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
 
Now we will use different types of Tokenizations 

* $BERT−Tokenizer$
* $Flan$ $T5$ $Base$ $V4$
* $Falcon$ $7B$
* $RoBERTa$

In [6]:
import numpy as np
from string import Template
import tqdm
import torch

from sklearn.decomposition import PCA

from transformers import (BertTokenizer, BertModel , 
                         T5Tokenizer, T5ForConditionalGeneration , 
                         AutoModelForCausalLM, BitsAndBytesConfig, AutoTokenizer , 
                         AutoModel)

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: l

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $3.1$ $|$ $BERT$ $Tokenizer$
$BERT$ $Bidirectional$ $Encoder$ $Representations$$ from $Transformers$ is a popular $Natural$ $Language$ $Processing$ $Model$ that has been shown to be `effective` for a variety of tasks, including 

* $Question$ $Answering$
* $Sentiment$ $Analysis$
* $Natural$ $Language$ $Inference$

One of the `key components` of BERT is the `tokenizer`, which is responsible for dividing text into tokens that can be understood by the model.

The $BERT$ $Tokenizer$ uses a technique called $Wordpiece$ $Tokenization$, which `breaks words` into `subwords`. This is done to address the `problem of rare words`, which may not be present in the BERT vocabulary. For example, the word "sleeping" would be tokenized as "sleep" and ""##ing".

In addition to wordpiece tokenization, the BERT tokenizer also adds a few special tokens to the vocabulary. These tokens include:

* `[CLS]` - This token is used to represent the beginning of a sentence.
* `[SEP]` - This token is used to represent the end of a sentence.
* `[PAD]` - This token is used to pad sequences to a fixed length.

In [7]:
bert_tokenizer = BertTokenizer.from_pretrained('bert-base-cased')

In [8]:
bert_tokenizer

BertTokenizer(name_or_path='bert-base-cased', vocab_size=28996, model_max_length=512, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True)

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

If we sent a sample text like `Why do you keep coming, every night in these dreams. Slowly also come in my arms one day`

In [9]:
bert_tokenizer("Why do you keep coming, every night in these dreams. Slowly also come in my arms one day" , return_tensors = "pt")

{'input_ids': tensor([[  101,  2009,  1202,  1128,  1712,  1909,   117,  1451,  1480,  1107,
          1292,  6149,   119, 13060,  1145,  1435,  1107,  1139,  1739,  1141,
          1285,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

We get the tokens which we can further use to get embedments

Lets now get the tokens for all train values, so we can predict the things better

In [10]:
def tokenize(value , tokenizer, index = "input_ids"):
    
    if len(value) > tokenizer.model_max_length:value = value[:tokenizer.model_max_length]
        
    tokens = tokenizer(value , return_tensors = "pt")

    return tokens[index]

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
 
This is our training data 

In [11]:
train_sum

student_id prompt_id  \
0     000e8c3c7ddb    814d6b   
1     0020ae56ffbf    ebad26   
2     004e978e639e    3b9047   
3     005ab0199905    3b9047   
4     0070c9e7af47    814d6b   
...            ...       ...   
7160  ff7c7e70df07    ebad26   
7161  ffc34d056498    3b9047   
7162  ffd1576d2e1b    3b9047   
7163  ffe4a98093b2    39c16e   
7164  fffbccfd8a08    ebad26   

                                                   text   content   wording  
0     The third wave was an experimentto see how peo...  0.205683  0.380538  
1     They would rub it up with soda to make the sme... -0.548304  0.506755  
2     In Egypt, there were many occupations and soci...  3.128928  4.231226  
3     The highest class was Pharaohs these people we... -0.210614 -0.471415  
4     The Third Wave developed  rapidly because the ...  3.272894  3.219757  
...                                                 ...       ...       ...  
7160  They used all sorts of chemical concoctions to...  0.205683  0.380538  
7161  The lowest classes are slaves and farmers slav... -0.308448  0.048171  
7162             they sorta made people start workin... -1.408180 -0.493603  
7163  An ideal tragety has three elements that make ... -0.393310  0.627128  
7164  The meat would smell sour but the would "rub i...  1.771596  0.547742  

[7165 rows x 5 columns]

In [12]:
train = train_pro.merge(train_sum , on = "prompt_id")
train.head()

prompt_id                                    prompt_question prompt_title  \
0    39c16e  Summarize at least 3 elements of an ideal trag...   On Tragedy   
1    39c16e  Summarize at least 3 elements of an ideal trag...   On Tragedy   
2    39c16e  Summarize at least 3 elements of an ideal trag...   On Tragedy   
3    39c16e  Summarize at least 3 elements of an ideal trag...   On Tragedy   
4    39c16e  Summarize at least 3 elements of an ideal trag...   On Tragedy   

                                         prompt_text    student_id  \
0  Chapter 13 \r\nAs the sequel to what has alrea...  00791789cc1f   
1  Chapter 13 \r\nAs the sequel to what has alrea...  0086ef22de8f   
2  Chapter 13 \r\nAs the sequel to what has alrea...  0094589c7a22   
3  Chapter 13 \r\nAs the sequel to what has alrea...  00cd5736026a   
4  Chapter 13 \r\nAs the sequel to what has alrea...  00d98b8ff756   

                                                text   content   wording  
0  1 element of an ideal tragedy is that it shoul... -0.210614 -0.471415  
1  The three elements of an ideal tragedy are:  H... -0.970237 -0.417058  
2  Aristotle states that an ideal tragedy should ... -0.387791 -0.584181  
3  One element of an Ideal tragedy is having a co...  0.088882 -0.594710  
4  The 3 ideal of tragedy is how complex you need... -0.687288 -0.460886

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
 
We do not need `prompt_id` amd `student_id`

In [13]:
train.drop(["prompt_id" , "student_id"] , axis = 1 , inplace = True)

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

Now we will tokenize every column

In [14]:
for columns in ["prompt_question" , "prompt_title" , "prompt_text" , "text"]:
    for index in tqdm.tqdm(range(train.shape[0]) , total = train.shape[0]):
        train[columns][index] = tokenize(train[columns][index] , tokenizer = bert_tokenizer , index = "input_ids")

  0%|          | 0/7165 [00:00<?, ?it/s]/tmp/ipykernel_21/218401772.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train[columns][index] = tokenize(train[columns][index] , tokenizer = bert_tokenizer , index = "input_ids")
100%|██████████| 7165/7165 [00:12<00:00, 593.80it/s]


In [15]:
train.head()

prompt_question  \
0  [[tensor(101), tensor(15463), tensor(12917), t...   
1  [[tensor(101), tensor(15463), tensor(12917), t...   
2  [[tensor(101), tensor(15463), tensor(12917), t...   
3  [[tensor(101), tensor(15463), tensor(12917), t...   
4  [[tensor(101), tensor(15463), tensor(12917), t...   

                                        prompt_title  \
0  [[tensor(101), tensor(1212), tensor(157), tens...   
1  [[tensor(101), tensor(1212), tensor(157), tens...   
2  [[tensor(101), tensor(1212), tensor(157), tens...   
3  [[tensor(101), tensor(1212), tensor(157), tens...   
4  [[tensor(101), tensor(1212), tensor(157), tens...   

                                         prompt_text  \
0  [[tensor(101), tensor(2943), tensor(1492), ten...   
1  [[tensor(101), tensor(2943), tensor(1492), ten...   
2  [[tensor(101), tensor(2943), tensor(1492), ten...   
3  [[tensor(101), tensor(2943), tensor(1492), ten...   
4  [[tensor(101), tensor(2943), tensor(1492), ten...   

                                                text   content   wording  
0  [[tensor(101), tensor(122), tensor(5290), tens... -0.210614 -0.471415  
1  [[tensor(101), tensor(1109), tensor(1210), ten... -0.970237 -0.417058  
2  [[tensor(101), tensor(18727), tensor(2231), te... -0.387791 -0.584181  
3  [[tensor(101), tensor(1448), tensor(5290), ten...  0.088882 -0.594710  
4  [[tensor(101), tensor(1109), tensor(124), tens... -0.687288 -0.460886

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
 
Lets now save this file 

In [16]:
os.makedirs("/kaggle/working/Bert Based")

In [17]:
train.to_csv("/kaggle/working/Bert Based/Train Embeds")

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $3.2$ $|$ $Flan$ $T5$ $Base$ $V4$

$Flan$ $T5$ $Base$ $V4$ is a $Large$ $Language$ $Model$ $LLM$ that was released by `Google AI` in `March 2023`. It is based on the `T5 Model`, which is a `text-to-text` `transfer transformer` model that was first introduced in $2020$. $Flan$ $T5$ $Base$ $V4$ has been `fine-tuned` on a mixture of tasks, including 
* $Text$ $Summarization$
* $Question$ $Answering$
* $Natural$ $Language$ $Inference$

$Flan$ $T5$ $Base$ $V4$ has $1.06$ `Million Parameters`, which is `smaller` than some other LLMs, such as $GPT-3$ and $Jurassic-1$ $Jumbo$. However, $Flan$ $T5$ $Base$ $V4$ has been shown to be `effective` for a variety of tasks, and it is relatively efficient to train and use.

In [18]:
flan_t5_tokenizer = T5Tokenizer.from_pretrained('/kaggle/input/flan-t5/pytorch/base/4')

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

If we pass a sample text like this

In [19]:
train = train_pro.merge(train_sum , on = "prompt_id")

In [20]:
stri = ""

for value in ["prompt_question" , "prompt_title" , "prompt_text" , "text"]:
    stri += "\n\n" + train[value][0]
    
print(stri) 



Summarize at least 3 elements of an ideal tragedy, as described by Aristotle.

On Tragedy

Chapter 13 
As the sequel to what has already been said, we must proceed to consider what the poet should aim at, and what he should avoid, in constructing his plots; and by what means the specific effect of Tragedy will be produced. 
A perfect tragedy should, as we have seen, be arranged not on the simple but on the complex plan. It should, moreover, imitate actions which excite pity and fear, this being the distinctive mark of tragic imitation. It follows plainly, in the first place, that the change of fortune presented must not be the spectacle of a virtuous man brought from prosperity to adversity: for this moves neither pity nor fear; it merely shocks us. Nor, again, that of a bad man passing from adversity to prosperity: for nothing can be more alien to the spirit of Tragedy; it possesses no single tragic quality; it neither satisfies the moral sense nor calls forth pity or fear. Nor, aga

In [21]:
flan_t5_tokenizer(stri , return_tensors = "pt")

Token indices sequence length is longer than the specified maximum sequence length for this model (944 > 512). Running this sequence through the model will result in indexing errors


{'input_ids': tensor([[12198,  1635,  1737,    44,   709,   220,  2479,    13,    46,  1523,
         18914,     6,    38,  3028,    57,    71, 17149,    17,   109,     5,
           461,  3083,  5402,    63,  8647,  1179,   282,     8, 19826,    12,
           125,    65,   641,   118,   243,     6,    62,   398,  8669,    12,
          1099,   125,     8, 11603,   225,  2674,    44,     6,    11,   125,
             3,    88,   225,  1792,     6,    16,     3, 24246,   112,  5944,
             7,   117,    11,    57,   125,   598,     8,   806,  1504,    13,
          3083,  5402,    63,    56,    36,  2546,     5,    71,   626, 18914,
           225,     6,    38,    62,    43,   894,     6,    36,     3,  9741,
            59,    30,     8,   650,    68,    30,     8,  1561,   515,     5,
            94,   225,     6,    72,  1890,     6,   256,  1734,  2874,    84,
          1215,  8464,     3,   102,   485,    11,  2971,     6,    48,   271,
             8, 11562,  3946,    13, 1

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

We can use our previous function and get the results

In [22]:
tokens = []
for index in tqdm.tqdm(range(train.shape[0]) , total = train.shape[0]):
    stri = ""
    for columns in ["prompt_question" , "prompt_title" , "prompt_text" , "text"]:
        stri += "\n\n" + str(train[columns][index])
        
    token = tokenize(train[columns][index] , tokenizer = flan_t5_tokenizer , index = "input_ids")
    tok = [x.detach().numpy().tolist() for x in token]
    
    tokens.append(tok)

100%|██████████| 7165/7165 [00:04<00:00, 1585.16it/s]


<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

Lets save a npy file for this tokenizer

In [23]:
os.makedirs("/kaggle/working/Flan T5")
np.save("/kaggle/working/Flan T5/Train Embeds" , np.array(tokens))

/tmp/ipykernel_21/1908650028.py:2: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  np.save("/kaggle/working/Flan T5/Train Embeds" , np.array(tokens))


<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $3.3$ $|$ $Falcon$ $7B$

The $Falcon$ $7B$ $Tokenizer$ is a tool that is used to prepare `text for input` to the $Falcon$ $7B$ $Language$ $Model$. It is based on the $WordPiece$ $Tokenization$ $Scheme$, which breaks words into subwords. This is done to address the problem of rare words, which may not be present in the Falcon 7B vocabulary. The $Falcon$ $7B$ $Tokenizer$ also adds a few special tokens to the vocabulary. These tokens include:

* `[CLS]` - This token is used to represent the beginning of a sentence.
* `[SEP]` - This token is used to represent the end of a sentence.
* `[PAD]` - This token is used to pad sequences to a fixed length.

It can be used for 

* $Text$ $Preparation$ for the $Falcon$ $7B$ $Language$ $Model$
* $Text$ $Summarization$
* $Question$ $Answering$
* $Natural$ $Language$ $Inference$
* $Code$ $Generation$
* $Translation$
* $Creative$ $Text$ $Generation$

In [24]:
falcon_tokenizer = AutoTokenizer.from_pretrained("tiiuae/falcon-7b")
falcon_tokenizer

PreTrainedTokenizerFast(name_or_path='tiiuae/falcon-7b', vocab_size=65024, model_max_length=2048, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'eos_token': '<|endoftext|>', 'additional_special_tokens': ['>>TITLE<<', '>>ABSTRACT<<', '>>INTRODUCTION<<', '>>SUMMARY<<', '>>COMMENT<<', '>>ANSWER<<', '>>QUESTION<<', '>>DOMAIN<<', '>>PREFIX<<', '>>SUFFIX<<', '>>MIDDLE<<']}, clean_up_tokenization_spaces=True)

In [25]:
falcon_tokenizer("Yayy")

{'input_ids': [49366, 100], 'token_type_ids': [0, 0], 'attention_mask': [1, 1]}

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

Now we will do the same thing we did for Flan T5

In [26]:
tokens = []
for index in tqdm.tqdm(range(train.shape[0]) , total = train.shape[0]):
    stri = ""
    for columns in ["prompt_question" , "prompt_title" , "prompt_text" , "text"]:
        stri += "\n\n" + str(train[columns][index])
        
    token = tokenize(train[columns][index] , tokenizer = falcon_tokenizer , index = "input_ids")
    tok = [x.detach().numpy().tolist() for x in token]
    
    tokens.append(tok)

100%|██████████| 7165/7165 [00:03<00:00, 1812.97it/s]


In [27]:
os.makedirs("/kaggle/working/Falcon 7B")
np.save("/kaggle/working/Falcon 7B/Train Embeds" , np.array(tokens))

/tmp/ipykernel_21/2071466847.py:2: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  np.save("/kaggle/working/Falcon 7B/Train Embeds" , np.array(tokens))


<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

## $3.4$ $|$ $RoBERTa$
    
$RoBERTa$ $Robustly$ $Optimized$ $BERT$ $Pretraining$ $Approach$ is a $Natural$ $Language$ $Processing$ $(NLP)$ model that was proposed in $2019$ by `Yinhan` Liu et al. It is a `reimplementation` of $BERT$ ($Bidirectional$ $Encoder$ $Representations$ from $Transformers$) with some `modifications` to the key `hyperparameters` and `minor embedding tweaks`. These modifications led to `significant performance gains` on a number of NLP tasks. $RoBERTa$ is based on the `transformer architecture`, which is a `Neural Network Architecture` that is particularly well-suited for NLP tasks. The transformer architecture uses `self-attention` to learn `long-range dependencies` between words in a sentence. This allows $RoBERTa$ to learn more `contextual representations` of words, which is important for many NLP tasks.

$RoBERTa$ is trained on a `massive dataset` of text and code. The dataset consists of `books`/`articles`/`code`. The dataset is `preprocessed` using `Byte-Level` `BPE` `(Byte Pair Encoding)`, which is a technique for tokenizing text into smaller units.

$RoBERTa$ is trained using a `Masked Language Modeling` ($MLM$) objective. In the MLM objective, some of the words in a `sentence are masked`, and the model is then trained to `predict the masked words`. This helps the model to `learn the contextual representations` of words.

In [28]:
roberta_tokenizer = AutoTokenizer.from_pretrained("roberta-base")

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

If we pass a sample text like `Gender was the concept invented by Bathroom Companies to sell more bathrooms`

In [29]:
tokens = roberta_tokenizer("Gender was the concept invented by Bathroom Companies to sell more bathrooms" , return_tensors = "pt")

In [30]:
roberta = AutoModel.from_pretrained("roberta-base")

Some weights of the model checkpoint at roberta-base were not used when initializing RobertaModel: ['lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

And pass it into the model

In [31]:
roberta(tokens["input_ids"])

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[-2.1888e-02,  6.7077e-02, -1.3544e-02,  ..., -9.9792e-02,
          -4.6635e-02, -2.2394e-02],
         [ 8.7412e-02,  8.6635e-02,  3.4988e-02,  ...,  3.3946e-02,
          -6.1401e-02, -1.3111e-01],
         [ 5.4624e-02,  1.2257e-01, -1.1354e-01,  ..., -8.6543e-02,
          -1.0259e-01,  9.4824e-05],
         ...,
         [ 4.4242e-02, -3.5742e-02,  1.3293e-01,  ..., -2.9737e-01,
          -2.6582e-02, -6.4403e-02],
         [ 4.7810e-03,  2.1526e-02, -9.3916e-02,  ...,  1.1625e-01,
           1.3474e-02, -8.4416e-02],
         [-5.3779e-03,  6.0121e-02, -3.9169e-02,  ..., -1.4106e-01,
          -4.4145e-02, -5.1654e-02]]], grad_fn=<NativeLayerNormBackward0>), pooler_output=tensor([[-5.8563e-02,  1.1118e-01,  2.0668e-02, -1.6277e-01,  2.3668e-01,
          2.6476e-01,  1.7605e-01, -2.2297e-01, -8.2392e-02, -7.8825e-02,
          3.1927e-01, -1.1081e-01, -2.1884e-01,  1.6666e-03,  2.1968e-01,
         -1.0258e-

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

We get a very long output, but we only need the $1^{st}$ index

In [32]:
roberta(tokens["input_ids"])[0] , roberta(tokens["input_ids"])[0].shape

(tensor([[[-2.1888e-02,  6.7077e-02, -1.3544e-02,  ..., -9.9792e-02,
           -4.6635e-02, -2.2394e-02],
          [ 8.7412e-02,  8.6635e-02,  3.4988e-02,  ...,  3.3946e-02,
           -6.1401e-02, -1.3111e-01],
          [ 5.4624e-02,  1.2257e-01, -1.1354e-01,  ..., -8.6543e-02,
           -1.0259e-01,  9.4824e-05],
          ...,
          [ 4.4242e-02, -3.5742e-02,  1.3293e-01,  ..., -2.9737e-01,
           -2.6582e-02, -6.4403e-02],
          [ 4.7810e-03,  2.1526e-02, -9.3916e-02,  ...,  1.1625e-01,
            1.3474e-02, -8.4416e-02],
          [-5.3779e-03,  6.0121e-02, -3.9169e-02,  ..., -1.4106e-01,
           -4.4145e-02, -5.1654e-02]]], grad_fn=<NativeLayerNormBackward0>),
 torch.Size([1, 15, 768]))

<div style="border-radius:10px; border:#800080 solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">

As we can see this is of shape $(15 , 768)$. 

These are all similar values and thus we can make them at one axis using $PCA$
    
I am not using `Kaggle GPUs` at this moment. I have imported the Embeddings from colab. Here is the code I used. The lengths are truncated to 512 for this moment 

```
import numpy as np 
import tqdm
from sklearn.decomposition import PCA
from transformers import AutoTokenizer , AutoModel

train_pro = pd.read_csv("/content/Original/prompts_train.csv")
train_sum = pd.read_csv("/content/Original/summaries_train.csv")
train = train_pro.merge(train_sum , on = "prompt_id")

tokenizer = AutoTokenizer.from_pretrained("roberta-base")
model = AutoModel.from_pretrained("roberta-base").to("cuda")

pca = PCA(n_components = 1)

embeds = []

for index in tqdm.tqdm(range(train.shape[0]) , total = train.shape[0]):

    torch.cuda.empty_cache()
    
    stri = ""
    
    for columns in ["prompt_question" , "prompt_title" , "prompt_text" , "text"]: stri += "\n\n" + str(train[columns][index])
    
    if len(stri) > tokenizer.model_max_length: stri = stri[:tokenizer.model_max_length]
    else:stri += " " * (tokenizer.model_max_length - len(stri))

    tokens = tokenizer(stri , return_tensors = "pt")["input_ids"].to("cuda")
    
    with torch.no_grad():x = model(input_ids=tokens)[0]
 
    x = pca.fit_transform(x.cpu().detach().numpy().squeeze().T).squeeze()

    torch.cuda.empty_cache()

    embeds.append(x)
```

In [33]:
os.makedirs("/kaggle/working/RoBERTa")
np.save("Train Embeds" , np.load("/kaggle/input/roberta/Roberta Train Embeds.npy"))

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#FFC0CB; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #FFC0CB">4 | TO DO LIST 📝</p>

<div style="border-radius:10px; border:#FFC0CB solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
<img src = "https://i.imgflip.com/7sgop6.jpg" width = 400>   

* TO DO 1 : EXPLORE THE DATA 
* TO DO 2 : VISUALIZE THE DATA 
* TO DO 3 : MAKE A PYTORCH DATALOADER
* TO DO 4 : MAKE A TENSORFLOW DATALOADER
* TO DO 5 : MAKE SOME TOKENIZERS
* TO DO 6 : TRAIN A PYTORCH MODEL
* TO DO 7 : TRAIN A TENSORFLOW MODEL
* TO DO 8 : IMRPOVE THE RESULTS
* TO DO 9 : LESS TRAINING TIME
* TO DO 10 : DANCE 

# <p style="font-family:JetBrains Mono; font-weight:bold; letter-spacing: 2px; color:#00FFFF; font-size:140%; text-align:left;padding: 0px; border-bottom: 3px solid #00FFFF">5 | Ending 🏁</p>

<div style="border-radius:10px; border:#00FFFF solid; padding: 15px; background-color: #F3f9ed; font-size:100%; text-align:left">
    
**THAT IT FOR TODAY GUYS**

**WE WILL GO DEEPER INTO THE DATA IN THE UPCOMING VERSIONS**

**PLEASE COMMENT YOUR THOUGHTS, HIHGLY APPRICIATED**

**DONT FORGET TO MAKE AN UPVOTE, IF YOU LIKED MY WORK $:)$**
    
<img src = "htTps://i.imgflip.com/19aadg.jpg">
    
**PEACE OUT $!!!$**